### PR TITLE
Update WatershedOperation

### PR DESCRIPTION
**Pull request by SamCarlberg**
Update WatershedOperation to return contours for the features discovered by watershed, instead of (a rather unuseful) Mat.

### CODE REVIEW

File: core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java [hunk 1]
```diff
@@ -35,11 +53,15 @@
             .initialValueSupplier(ContoursReport::new)
             .build();
 
-    private final SocketHint<Mat> outputHint = SocketHints.Inputs.createMatSocketHint("Output", true);
+    @SuppressWarnings("unchecked")
+    private final SocketHint<ContoursReport> outputHint = new SocketHint.Builder<>(ContoursReport.class)
+            .identifier("Segmented Contours")
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: This is a bit too long; it gets cut off and rendered as "`Segmented C...`" on my machine. Maybe just "`Features`"?

**Comment by JLLeitschuh** [comment]
You're going to need to rebase this so that it is just the one commit. Currently the diff is to large to correctly see what has been changed.

**Comment by SamCarlberg** [comment]
Wait what, it was rebased. Weird.

File: core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java [hunk 2]
```diff
@@ -91,19 +103,51 @@ public void perform() {
...
+      int bufIdx;
+      for (int y = 0; y < output.rows(); y++) {
+        for (int x = 0; x < output.cols(); x++) {
+          bufIdx = (stride * y) + (x * markers.channels());
+          byte val = buffer.get(bufIdx);
+          if (val == 0 || val == -1) {
+            // Background (0) or border (-1)
+            continue;
+          }
+          Mat m = segments.computeIfAbsent(val, k -> {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: When does this mat get used?

Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: It's a temporary Mat used to find individual contours from the watershed.

**Comment by SamCarlberg** [comment]
Crap, still getting segfaults. Only seems to happen when:

1. There are a ton of contours (> 100); and
2. The `FindContours` operation doesn't use `External only`

Tons of external-only contours either works normally or throws an exception (if `num_contours > 253`).  
Fewer non-`External only` contours also throws an exception. That gets caught [here](https://github.com/WPIRoboticsProjects/GRIP/pull/607/files#diff-452222b1cce028acf7b1292ab7bea63fR151), which _should_ also catch the conditions for the segfaults before they happen

File: core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java [hunk 3]
```diff
@@ -85,29 +97,85 @@ public void perform() {
...
+      }
+    }
+    if (!found) {
+      // Should only happen if the image is clogged with contours
+      throw new IllegalStateException("Could not find a point for the background label");
+    }
+    return backgroundLabel;
+  }
+
+  private Point fromPoint2f(Point2f p) {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: Make this static?

File: core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java [hunk 4]
```diff
@@ -85,29 +97,85 @@ public void perform() {
...
     } finally {
       // make sure that the working mat is freed to avoid a memory leak
       markers.release();
     }
   }
 
+  /**
+   * Finds the first available point to place a background marker for the watershed operation.
+   */
+  private Point2f findBackgroundMarker(Mat markers, MatVector contours) {
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: Can this be static?

Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: It could, but there's no real need.

Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: It doesn't use member variables so it should be static.

File: core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java [hunk 5]
```diff
@@ -85,29 +97,85 @@ public void perform() {
...
   }
 
+  /**
+   * Finds the first available point to place a background marker for the watershed operation.
+   */
+  private static Point2f findBackgroundMarker(Mat markers, MatVector contours) {
+    final int cols = markers.cols();
+    final int rows = markers.rows();
+    final int minDist = 5;
+    Point2f backgroundLabel = new Point2f();
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: Is this a memory resource that should be closed once done with it?

Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/WatershedOperation.java]: There's no need to manually free it, the GC will get it.

**Comment by JLLeitschuh** [comment]
LGTM! We'll need to mention that this is a breaking change in the release notes.

**Pull request merged**